### ISSUE TITLE
(bug) Integration tests fail on fresh install of debian/mariadb

### ISSUE THREAD

**Issue by jniles** [third-culture-software/bhima#1831]
**Operating System**: Debian 9
**MySQL/MariaDB Server Version**: 10.1.23-MariaDB-9+deb9u1
**Node Version**: 8.1.3
**NPM Version**: 5.0.3

On a fresh install in a VM, the integration tests for the project fail with the mysql error `ER_TRUNCATED_WRONG_VALUE`.  There may be more errors.  Since our tests are all linked, the cascade of failures affects 35 tests, but it is unclear how many are directly caused by MySQL command failures and how many are simply expecting data that hasn't been built.

**Checklist**
 - [x] sql-mode="STRICT_ALL_TABLES"
 - [x] Reproducible on client using a caution payment and test dataset.
 - [x] Reproducible every run (non-random)
 
When manually testing, the following log is surfaced:
```sh
debug: [Transaction] CALL StageCash(12, 2, 2, X'3be232f9a4b94af6984c5d3f87d5c107', 1, '2017-07-09 17:08:08.873', 1, 0, 'Cash payment by Test 2 Patient (PA.TPA.2) against invoice(s) IV.TPA.1. -- ', X'61f0cd0386f24efb815b399669c6ca8e')
debug: [Transaction] CALL StageCashItem(X'5855131d1dff44ec86b6878a4d1e827a', X'61f0cd0386f24efb815b399669c6ca8e', X'c44619e03a884754a750a414fc9567bf')
debug: [Transaction] CALL CalculateCashInvoiceBalances(X'61f0cd0386f24efb815b399669c6ca8e')
debug: [Transaction] CALL WriteCash(X'61f0cd0386f24efb815b399669c6ca8e')
debug: [Transaction] CALL WriteCashItems(X'61f0cd0386f24efb815b399669c6ca8e')
debug: [Transaction] CALL PostCash(X'61f0cd0386f24efb815b399669c6ca8e')
debug:  Error: ER_TRUNCATED_WRONG_VALUE: Truncated incorrect DOUBLE value: 'NS2'
    at Query.Sequence._packetToError (/home/jniles/staging-server/node_modules/mysql/lib/protocol/sequences/Sequence.js:52:14)
    at Query.ErrorPacket (/home/jniles/staging-server/node_modules/mysql/lib/protocol/sequences/Query.js:77:18)
    at Protocol._parsePacket (/home/jniles/staging-server/node_modules/mysql/lib/protocol/Protocol.js:280:23)
    at Parser.write (/home/jniles/staging-server/node_modules/mysql/lib/protocol/Parser.js:75:12)
    at Protocol.write (/home/jniles/staging-server/node_modules/mysql/lib/protocol/Protocol.js:39:16)
    at Socket.<anonymous> (/home/jniles/staging-server/node_modules/mysql/lib/Connection.js:103:28)
    at emitOne (events.js:115:13)
    at Socket.emit (events.js:210:7)
    at addChunk (_stream_readable.js:252:12)
    at readableAddChunk (_stream_readable.js:239:11)
    at Socket.Readable.push (_stream_readable.js:197:10)
    at TCP.onread (net.js:589:20)
    --------------------
    at Protocol._enqueue (/home/jniles/staging-server/node_modules/mysql/lib/protocol/Protocol.js:141:48)
    at PoolConnection.query (/home/jniles/staging-server/node_modules/mysql/lib/Connection.js:208:25)
    at queryConnection (/home/jniles/staging-server/bin/server/lib/db/transaction.js:14:28)
    at queries.map.bundle (/home/jniles/staging-server/bin/server/lib/db/transaction.js:118:21)
    at Array.map (native)
    at Query.connection.beginTransaction [as _callback] (/home/jniles/staging-server/bin/server/lib/db/transaction.js:117:34)
    at Query.Sequence.end (/home/jniles/staging-server/node_modules/mysql/lib/protocol/sequences/Sequence.js:86:24)
    at Query._handleFinalResultPacket (/home/jniles/staging-server/node_modules/mysql/lib/protocol/sequences/Query.js:137:8)
    at Query.OkPacket (/home/jniles/staging-server/node_modules/mysql/lib/protocol/sequences/Query.js:72:10)
    at Protocol._parsePacket (/home/jniles/staging-server/node_modules/mysql/lib/protocol/Protocol.js:280:23)
    at Parser.write (/home/jniles/staging-server/node_modules/mysql/lib/protocol/Parser.js:75:12)
    at Protocol.write (/home/jniles/staging-server/node_modules/mysql/lib/protocol/Protocol.js:39:16)
    at Socket.<anonymous> (/home/jniles/staging-server/node_modules/mysql/lib/Connection.js:103:28)
    at emitOne (events.js:115:13)
    at Socket.emit (events.js:210:7)
    at addChunk (_stream_readable.js:252:12)
    at readableAddChunk (_stream_readable.js:239:11)
error: [Transaction] Encountered error ER_TRUNCATED_WRONG_VALUE. Rolling transaction back and recoverying database connections.
```

In my mind, this is an issue with the `GenerateTransId()` function.  It looks like `TRANS2` is a truncated to `NS2` and attempted to get put into the `trans_id` field (which only takes numeric values).  It is either an issue with the truncation logic or with our test dataset not being correct.

**Steps to Reproduce**
 1. Set up a virtual machine following the instructions in the [INSTALL](https://github.com/IMA-WorldHealth/bhima-2.X/blob/master/docs/INSTALL.md) file.  Ensure the versions of dependencies match the above versions.
 2. Run `npm run test:integration`.

**Comment by jniles** [comment]
I have located the source of this bug.  The problem is that we do not calculate the length of the project abbreviation field, and if the abbreviation changes length, the system no longer knows how to increment the project identifier.

In the test data, the project abbreviation changes for project A.  In the `posting_journal`, there are prefabricated transactions with trans_ids `TRANS1`.  However, in the `project` table, the abbreviation for project 1 is `TPA`.  When the MySQL function `GenerateTransactionId(1)` is called, it determines that the correct project abbreviation length to look for is 4, thus truncating all the `TRANS*` transaction identifiers to `NS*` and finding the MAX().  It then tries to add 1 to the MAX which is a string + integer and attempts to write that to the database.  This is disallowed under STRICT_ALL_TABLES.

I'm unclear why this error does not show up in Travis CI.  I believe Travis is running MySQL Version 5.6 on an ubuntu server.  Neither of those should make a difference.

**Recommended Fix**
I propose that we simply make `TRANS*` in the test dataset (and all subsequent tests that use it) into `TPX*`.  That's the simplest solution available for the moment and shouldn't affect any other tests.

**Comment by jniles** [comment]
This was fixed accidentally in #2173
